File: go/day01/main.go
package main

import (
	"aoc2024/utils"
	"slices"
	"strconv"
	"strings"
)

func buildLists(path string) ([]int, []int) {
	input := utils.ReadInput(path)

	var list1 []int
	var list2 []int

	for _, line := range input {
		numberStrings := strings.Fields(line)
		number1, err := strconv.Atoi(numberStrings[0])

		if err != nil {
			panic(err)
		}

		number2, err := strconv.Atoi(numberStrings[1])

		if err != nil {
			panic(err)
		}

		list1 = append(list1, number1)
		list2 = append(list2, number2)
	}

	return list1, list2
}

func Part1(list1 []int, list2 []int) {
	slices.Sort(list1)
	slices.Sort(list2)

	difference := 0

	for i := 0; i < len(list1); i++ {
		difference += utils.Abs(list1[i] - list2[i])
	}

	println("Part 1: ", difference)
}

func Part2(list1 []int, list2 []int) {
	total := 0

	occurances := make(map[int]int)
	minIndex := 0

	for i := 0; i < len(list1); i++ {
		current := list1[i]

		if count, ok := occurances[current]; ok {
			total += count * current
		} else {
			occurances[current] = 0
			for j := minIndex; j < len(list2); j++ {
				if list2[j] == current {
					occurances[current]++
				}
				if list2[j] > current {
					minIndex = j
					break
				}
			}
			total += occurances[current] * current
		}

	}

	println("Part 2: ", total)
}

func main() {
	list1, list2 := buildLists("day01/input.txt")
	Part1(list1, list2)
	Part2(list1, list2)
}

File: rust/day01/src/main.rs
use std::collections::HashMap;

use utils;

fn build_lists(path: &str) -> (Vec<i32>, Vec<i32>) {
    let input = utils::read_input(path);
    
    input
        .lines()
        .map(|line| line
            .split_whitespace()
            .map(|x| x.parse::<i32>())
            .map(|x| x.unwrap())
            .collect::<Vec<i32>>()
        )
        .fold((Vec::new(), Vec::new()), |(mut list1, mut list2), x| {
            if x.len() == 2 {
                list1.push(x[0]);
                list2.push(x[1]);
            }
            (list1, list2)
        })
}

fn part1(list1: Vec<i32>, list2: Vec<i32>) {
    let mut difference: i32 = 0;

    for i in 0..list1.len() {
        difference += (list1[i] - list2[i]).abs();
    }

    println!("Part 1: {}", difference);
}

fn part2(list1: Vec<i32>, list2: Vec<i32>) {
    let mut total = 0;

    let mut occurances: HashMap<i32, i32> = HashMap::new();
    let mut min_index: usize = 0;

    for i in 0..list1.len() {
        let current = list1[i];

        if let Some(count) = occurances.get(&current) {
            total += current * count;
        } else {
            occurances.insert(current, 0);
            for j in min_index..list2.len() {
                if list2[j] == current {
                    let count = occurances.get_mut(&current).unwrap();
                    *count += 1;
                }
                if list2[j] > current {
                    min_index = j;
                    break;
                }
            }
            let count = occurances.get(&current).unwrap();
            total += current * count;
        }
    }

    println!("Part 2: {}", total);
}

fn main() {
    let (mut list1, mut list2) = build_lists("day01/input.txt");
    list1.sort();
    list2.sort();
    part1(list1.clone(), list2.clone());
    part2(list1, list2);
}
